File: Algorithms/0814_Binary_Tree_Pruning/Java/BinaryTreePruningSolution1.java

// Space: O(1)
// Time: O(n)

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public TreeNode pruneTree(TreeNode root) {
        if (root == null) return root;
        if (root.left == null && root.right == null) {
            return root.val == 1 ? root : null;
        }
        int res = dfs(root);
        return res == 0 ? null : root;
    }

    private int dfs(TreeNode root) {
        if (root == null) return 0;
        if (root.left == null && root.right == null) return root.val;

        int left = dfs(root.left);
        int right = dfs(root.right);
        if (left == 0) root.left = null;
        if (right == 0) root.right = null;

        if (left == 0 && right == 0 && root.val == 0) return 0;
        return 1;
    }
}

File: Algorithms/0814_Binary_Tree_Pruning/Python/Binary_Tree_Pruning_Solution_1.py

# Space: O(1)
# Time: O(n)

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def pruneTree(self, root):
        if root is None: return root
        if root.left is None and root.right is None:
            if root.val == 0: return None
            return root

        def dfs(root):
            if root is None: return 0
            if root.left is None and root.right is None: return root.val

            left = dfs(root.left)
            right = dfs(root.right)

            if left == 0: root.left = None
            if right == 0: root.right = None
            if left == 0 and right == 0 and root.val == 0: return 0
            return 1

        res = dfs(root)
        return root if res != 0 else None




